contracts/src/utils/Counter.sol

```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.20;
+
+library Counter {
+    function createCounter(uint256 length) internal pure returns (uint256[] memory counter) {
+        uint256 counterLength = length / 16 + (length % 16 == 0 ? 0 : 1);
+        counter = new uint256[](counterLength);
+    }
+
+    function get(uint256[] storage self, uint256 index) internal view returns (uint16) {
+        uint256 element = index >> 4;
+        uint8 inside = uint8(index) & 0x0F;
+        return uint16((self[element] >> (16 * inside)) & 0xFFFF);
+    }
+
+    function set(uint256[] storage self, uint256 index, uint16 value) internal {
+        uint256 element = index >> 4;
+        uint8 inside = uint8(index) & 0x0F;
+        uint256 zero = uint256(0xFFFF) << (16 * inside);
+        uint256 shiftedValue = uint256(value) << (16 * inside);
+        self[element] = self[element] & zero | shiftedValue;
+    }
```

Comment:
Seems nuance that makes `value = get(index)` not always the same value as set by `set(index, value)`, Is that as expected?

Reply:
This shouldn't be the case. I have added more unit tests for counters:
https://github.com/Snowfork/snowbridge/pull/956/commits/c631af0355189f8414064e4d9a8f32fea317f302

What inputs did you use to produce the behavior? I can add it as a test case.

Reply:
```
function testCounterGetAndSetNotMatch() public {
        counters = Counter.createCounter(300);
        uint256 index = 0;
        uint16 value = 11;
        counters.set(index, value);
        uint16 new_value = counters.get(index);
        console.log("round1:index at %d set %d and get %d", index, value, new_value);
        assertEq(value, new_value);
        value = value + 1;
        counters.set(index, value);
        new_value = counters.get(index);
        console.log("round2:index at %d set %d and get %d", index, value, new_value);
        assertEq(value, new_value);
}
```

Test failed at round2 assert.

Reply:
Nice catch! Addressed in [0e0978d](https://github.com/Snowfork/snowbridge/pull/956/commits/0e0978d8bef1aac762889f22b5c8f1d53e2fa286).